In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats

<h3><b>1. Expected Shortfall</b></h3>

<p>
ES = (ϕ(z) / (1 − α)) ⋅ σ
</p>

<ul>
  <li><b>ϕ(z)</b> = PDF at z</li>
  <li><b>α</b> = Confidence Interval</li>
  <li><b>σ</b> = Standard Deviation</li>
</ul>


In [ ]:
# Q1. Given a portfolio’s VaR at 95% confidence is 10,000, compute the Expected Shortfall (ES) if the tail loss distribution follows a normal distribution.
# ES: Average loss in worst cases
var = 10000
ci = 0.95

# Z-score
z = stats.norm.ppf(ci)

# Standard deviantion (From: VaR = mu + sigma * z)
sigma = var / z

# Probability density function at z score
phi_z = stats.norm.pdf(z)

# Espected Shortfall
ES = phi_z / (1 - ci) * sigma

print(f"Espected Shortfall: {ES:.2f}")

Espected Shortfall: 12540.40


**Conclusion**: The average loss in worst 5% of the cases is about 12540.40

<h3><b>2. Sharpe Ratio</b></h3>

<p>
Sharpe Ratio = (Expected Return − Risk-Free Rate) / σ
</p>

<ul>
  <li><b>Expected Return</b> = Portfolio's average return</li>
  <li><b>Risk-Free Rate</b> = Return of a risk-free asset</li>
  <li><b>σ</b> = Standard Deviation (risk)</li>
</ul>


In [ ]:
# Q2. A portfolio has an expected annual return of 12%, a risk-free rate of 2%, and a standard deviation of 18%. Compute the Sharpe Ratio.
# Sharpe Ratio: Calculates how much more return we get for each unit risk
expected_return = 0.12
risk_free_rate = 0.02
std = 0.18

# Sharpe Ratio
sharpe_ratio = (expected_return - risk_free_rate) / std

print(f"Sharpe Ratio: {sharpe_ratio:.2f}")

Sharpe Ratio: 0.56


**Conclusion**: A Sharpe Ratio of 0.56 means you're earning 0.56 units of excess return for each unit of risk taken

<h3><b>3. Monte Carlo Simulated Return</b></h3>

<p>
Simulated Return = μ &times; t + σ &times; √t &times; z
</p>

<ul>
  <li><b>μ</b> = Mean (expected) return</li>
  <li><b>σ</b> = Volatility (standard deviation of returns)</li>
  <li><b>t</b> = Time horizon (in years)</li>
  <li><b>z</b> = Random number from standard normal distribution (𝒩(0,1))</li>
</ul>

In [ ]:
# Q3. Given mean return = 7%, volatility = 12%, and time horizon = 1 year, simulate one possible future portfolio return using a Monte Carlo model.
# Monte Carlo Model: Helps by calculating future outcomes by randomness
np.random.seed(42)
mu = 0.07
sigma = 0.12
t = 1     #1 year time

z = np.random.randn()

simulated_return = mu * t + sigma * np.sqrt(t) * z
print(f"Normal Distribution: {z:.2f}")
print(f"Simulated Future Return: {simulated_return * 100:.2f}%")

Normal Distribution: 0.50
Simulated Future Return: 12.96%


**Conclusion**: Future returns in this case are 12.96% for z = 0.50. We can simulate different outcomes if the np.random.seed(42) is removed.